enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,197 @@
+* As an Openshift engineer, having nodes boot up on an unsupported OCP version is a security liability. By having nodes directly boot on the release payload image, it helps me avoid tracking incompatibilities across OCP release versions and shore up technical debt(see issues linked above). 
+
+* As a cluster administrator, having to keep track of a "boot" vs "live" image for a given cluster is not intuitive or user friendly. In the worst case scenario, I will have to reset a cluster(or do a lot of manual steps with rh-support in recovering the node) simply to be able to scale up nodes after an upgrade. If I'm managing an IPI cluster, once opted in, this feature will be a "switch on and forget" mechanism for me. If I'm managing a UPI cluster, this would provide me with documentation that I could follow after an upgrade to ensure my cluster has the latest bootimages.
+
+### Goals
+
+The MCO will take over management of the boot image references and the stub ignition. The installer is still responsible for creating the `MachineSet` at cluster bring-up of course, but once cluster installation is complete the MCO will ensure that boot images are in sync with the latest payload. From the user standpoint, this should cause less compatibility issues as nodes will no longer need to pivot to a different version of rhcos during node scaleup.
+
+### Non-Goals
+
+- The new subcontroller does not provide a solution for UPI as it does not use `MachineSets`. We plan to support a UPI solution via documentation that is based on this workflow.
+- This is meant to be a user opt-in feature, and if the user wishes to keep their boot images static it will let them do so.
+- This does not intend to solve [booting into custom pools](https://issues.redhat.com/browse/MCO-773). 
+
+## Proposal
+
+This automated flow is fairly straightforward, but will require a bit of special casing for each platform. 
+
+- The `machine-config-controller`(MCC) pod will gain a new sub-controller `machine_set_controller`(MSC) that monitors `MachineSet` changes and the `coreos-bootimages` [ConfigMap](https://github.com/openshift/installer/pull/4760).
+- Based on platform and arch type, the MSC will check if the images referenced in the `MachineSet(s)` is the same as the one in the ConfigMap. Each platform(gcp, aws...and so on) does this differently, so this is a good opportunity to split the work up between platforms and see if the implementation is effective. The ConfigMap is considered to be the golden set of bootimage values, i.e. they will never go out of date.
+- Next, it will check if the stub secret referenced is spec 3. If it is spec 2, the MSC will try create a new version of this secret by trying to translate it to spec 3. This step is platform/arch agnostic. Failure to up translate will cause a degrade and the sub-controller will exit without patching the `MachineSet`.
+- Finally, if the MSC will attempt to patch the `MachineSet` if required. Failure to do so will cause a degrade. 
```

Comment:
What's the intended ordering with MSC and in-place node updates?  Do they just default to run concurrently?

IOW right after I start `oc adm upgrade`, while the first control plane node and worker are being drained, are my machinesets already being patched?

Or do we e.g. wait until at least one node has successfully done an in-place update to the new version before patching machinesets?  One argument to do this is that it at least acts as a form of sanity checking.

Reply:
The current design would do it concurrently, yes. In my testing, it happened before the node updates take place, but.. my test runs were not very exhaustive :) This is probably worth calling out: in the  case of GCP - master nodes did not have machinesets, so they'd never get new boot images via this mechanism. Overall though, a sanity check is definitely worth considering, I'll refine on the strategy after giving it some thought!

Reply:
I gave this some thought - how do we feel about an annotation(say `io.openshift.mco-image-verified=true`)  to the golden configmap? It sort of makes sense to tie this to the configmap because we are affirming that the references stored within it have gone through at least one in place update.

So a proposed flow:
- When the configmap is updated by CVO, we remove this annotation if it exists. This is to indicate that an upgrade is in progress and to not trust the "new" configmap just yet. 
- The MSC will stop reconciling all machinesets as the configmap no longer has this annotation.
- Once atleast one node complete the update to the payload, some component within the MCO(TBD what, open to ideas) will now annotate the configmap with `io.openshift.mco-image-verified=true`
- The MSC will now begin to reconcile all the machinesets as normal, until opted-out/another upgrade takes place

Reply:
Hmm, so in this proposal, only are we in steady state do we use new bootimages?

I think there's merit to both sides of this coin. We can try implementing it this way to start since it should be a bit "safer". But note that any failures to this flow will be after the upgrade has completed, which may not be the best UX

Reply:
Discussed this with Jerry and we came to the conclusion that a simple true/false is not sufficient. We will need to verify that the bootimages configmap is on the same cluster version as the MCO, so we don't have rolling upgrade issues. I'll flesh this out more in the EP.

Reply:
I just noticed that there is a [release version field](https://github.com/openshift/installer/blob/master/hack/build-coreos-manifest.go#L52) in the golden configmap - perhaps that can be used somehow here, instead of an annotation based system? @cgwalters It looks like it is still hardcoded to be `0.0.1-snapshot`, I assume for initial testing purposes